pkg/controller/storageprofile-controller.go

```diff
@@ -0,0 +1,222 @@
+	// reconcile accessModes
+	if len(claimPropertySet.AccessModes) == 0 {
+		capabilities, found := storagecapabilities.Get(sc)
+		if found {
+			claimPropertySet.AccessModes = []v1.PersistentVolumeAccessMode{capabilities.AccessMode}
+		}
+	}
+}
+
+func (r *StorageProfileReconciler) createStorageProfile(sc *storagev1.StorageClass) (*cdiv1.StorageProfile, error) {
+	storageProfile := &cdiv1.StorageProfile{}
+	if err := r.uncachedClient.Get(context.TODO(), types.NamespacedName{Name: sc.Name}, storageProfile); err != nil {
+		if k8serrors.IsNotFound(err) {
+			storageProfile := MakeEmptyStorageProfileSpec(sc.Name)
+			if err := operator.SetOwnerRuntime(r.uncachedClient, storageProfile); err != nil {
+				return nil, err
+			}
+			if err := r.client.Create(context.TODO(), storageProfile); err != nil {
+				if k8serrors.IsAlreadyExists(err) {
+					storageProfile := &cdiv1.StorageProfile{}
+					if err := r.uncachedClient.Get(context.TODO(), types.NamespacedName{Name: sc.Name}, storageProfile); err == nil {
+						return storageProfile, nil
+					}
+					return nil, err
+				}
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
+	}
+	return storageProfile, nil
+}
+
+// MakeEmptyStorageProfileSpec creates StorageProfile manifest
+func MakeEmptyStorageProfileSpec(name string) *cdiv1.StorageProfile {
+	return &cdiv1.StorageProfile{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "StorageProfile",
+			APIVersion: "cdi.kubevirt.io/v1beta1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Labels: map[string]string{
+				common.CDILabelKey:       common.CDILabelValue,
+				common.CDIComponentLabel: "",
+			},
+		},
+	}
+}
+
+// NewStorageProfileController creates a new instance of the StorageProfile controller.
+func NewStorageProfileController(mgr manager.Manager, log logr.Logger) (controller.Controller, error) {
+	uncachedClient, err := client.New(mgr.GetConfig(), client.Options{
+		Scheme: mgr.GetScheme(),
+		Mapper: mgr.GetRESTMapper(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	reconciler := &StorageProfileReconciler{
+		client:         mgr.GetClient(),
+		uncachedClient: uncachedClient,
+		scheme:         mgr.GetScheme(),
+		log:            log.WithName("storageprofile-controller"),
+	}
+
+	storageProfileController, err := controller.New(
+		"storageprofile-controller",
+		mgr,
+		controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	if err := addStorageProfileControllerWatches(mgr, storageProfileController, log); err != nil {
+		return nil, err
+	}
+	if err := reconciler.Init(); err != nil {
+		log.Error(err, "Unable to initalize StorageProfile objects")
+	}
+
+	log.Info("Initialized StorageProfile objects")
+	return storageProfileController, nil
+}
+
+func addStorageProfileControllerWatches(mgr manager.Manager, c controller.Controller, log logr.Logger) error {
+	// Add schemes.
```

Comment:
what's the diff between the schemes? do we need to Watch both of them?

Reply:
the storagev1 contains functions and definitions for k8s storage  - storagev1.StorageClass, the cdiv1 defines definitions for CDI - cdiv1.StorageProfile.

Reply:
Good to learn!